lxd-agent/network.go

```diff
@@ -69,3 +78,101 @@ func serverTLSConfig() (*tls.Config, error) {
 
 	return tlsConfig, nil
 }
+
+// reconfigureNetworkInterfaces checks for the existence of "nics.json" in the config share.
+// Interfaces are identified by their MAC address and configured accordingly.
+func reconfigureNetworkInterfaces() {
+	// Abort if configuration file does not exist (nothing to do), otherwise log and return.
+	nicDir, err := ioutil.ReadDir("nics")
+	if os.IsNotExist(err) {
+		return
+	} else if err != nil {
+		logger.Errorf("Could not read network interface configuration directory: %s", err.Error())
+		return
+	}
+
+	nicData := make(map[string]deviceConfig.NICConfig)
+	for _, f := range nicDir {
+		nicBytes, err := ioutil.ReadFile(filepath.Join("nics", f.Name()))
+		if err != nil {
+			logger.Errorf("Could not read network interface configuration file: %s", err.Error())
+		}
+
+		var conf deviceConfig.NICConfig
+		err = json.Unmarshal(nicBytes, &conf)
+		if err != nil {
+			logger.Errorf("Could not parse network interface configuration file: %s", err.Error())
+			return
+		}
+
+		nicData[f.Name()] = conf
+	}
+
+	netDir := "/sys/class/net"
+	files, err := ioutil.ReadDir(netDir)
+	if err != nil {
+		return
+	}
+
+	reverter := revert.New()
+	for _, f := range files {
+		addrBytes, err := ioutil.ReadFile(filepath.Join(netDir, f.Name(), "address"))
+		if err != nil {
+			return
+		}
+		macAddress := string(bytes.TrimRight(addrBytes, "\n"))
+
+		mtuBytes, err := ioutil.ReadFile(filepath.Join(netDir, f.Name(), "mtu"))
+		if err != nil {
+			return
+		}
+		mtuStr := string(bytes.TrimRight(mtuBytes, "\n"))
+
+		nic, ok := nicData[macAddress]
+		if !ok {
+			continue
+		}
+
+		_, err = shared.RunCommand("ip", "link", "set", "down", f.Name())
+		if err != nil {
+			logger.Error("Could not set interface %q to down: %s", f.Name(), err.Error())
+			return
+		}
+		reverter.Add(func() {
+			shared.RunCommand("ip", "link", "set", "up", f.Name())
+		})
+
+		_, err = shared.RunCommand("ip", "link", "set", f.Name(), "name", nic.Name)
+		if err != nil {
+			reverter.Fail()
+			logger.Errorf("Could not rename interface %q to %q: %s", f.Name(), nic.Name, err.Error())
+			return
+		}
+		reverter.Add(func() {
+			_, err = shared.RunCommand("ip", "link", "set", nic.Name, "name", f.Name())
+		})
+
+		if nic.MTU != 0 {
+			_, err = shared.RunCommand("ip", "link", "set", "dev", nic.Name, "mtu", strconv.Itoa(int(nic.MTU)))
+			if err != nil {
+				reverter.Fail()
```

Comment:
If you find yourself calling `revert.Fail()` multiple times inside a function then you're likely using it wrong.

The usage pattern should be:

```go
func foo() error {
    revert := revert.New()
   defer revert.Fail()

   err := doSomething() 
   if err != nil {
      return err
   }
   revert.Add(func() { // Add revert to undo the thing you just did})

   err := doSomethingElse() 
   if err != nil {
      return err // I don't have to worry about calling revert.Fai() as its been deferred and now my revert steps will be run in reverse order.
   }
   revert.Add(func() { // Add revert to undo the thing you just did})

  // If we got here nothing went wrong (happy path) so cancel the reverter. 
  revert.Success()
  return nil
}